### PR TITLE
graph_trainer: runtime estimator (cost-model / benchmark / interpreter)

### PR DESCRIPTION
**Pull request by nnurlan008**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3828

## What this adds

A static/measured runtime estimator over the joint fwd+loss+bwd FX graph produced
by `minimal_fx_tracer`, for a future activation-policy (save / recompute /
offload) solver.

- `runtime_estimator.py` -- `RuntimeEstimator`, three modes (selected via
  `RuntimeEstimator()(mode)`), mirroring upstream
  `torch.distributed._tools.runtime_estimator.RuntimeEstimator`:
  - `operator-level-cost-model` -- static roofline, `max(compute, HBM transfer)`
    per FX node (no GPU execution).
  - `operator-level-benchmark` -- per-op real kernel time on random tensors
    (warmup + timed CUDA-event iters); each op benchmarked in isolation, so it
    works even for a model too big to run end-to-end.
  - `operator-level-interpreter` -- runs the real graph via `run_traced` and times
    it: real end-to-end total (overlap-aware) + a per-node CUDA-event breakdown.
  It also aggregates per-module forward/backward runtimes (`display_modulewise_stats`).
- `trainer.py` logs a `[runtime-estimation]` line each traced step, comparing the
  three modes on the real model.

## How to validate / reproduce (reviewer steps)

Run from the repo root. Setup:

    source ~/venvs/torchtitan/bin/activate
    # devvm only: fetch tokenizer/dataset over the proxy
    export https_proxy=http://fwdproxy:8080 http_proxy=http://fwdproxy:8080

cudagraph + the memory-policy/offload/remat passes are disabled with
`--compile.disable_passes` so the estimate runs on the baseline (save-everything)
graph. The estimator logs `[runtime-estimation]` on the POST-regional-inductor
graph -- read THAT one (the pre-regional graph runs eager flex, ~5x slower and
not representative). After "After regional inductor pass:", look for:

    [runtime-estimation] per-op-sum (ms): cost-model=..  benchmark=..  interpreter=.. |
      interpreter end-to-end=.. (overlap) | closeness vs benchmark: .. | vs interpreter end-to-end: ..

Note: the `graph_trainer_llama3_1b` / `graph_trainer_qwen3_1_7b` configs used
below are not included in this PR; add the corresponding model
`config_registry.py` entries locally to reproduce.

Single config (vary `--training.local_batch_size` / `--training.seq_len`):

    NGPU=1 MODULE=graph_trainer.llama3 CONFIG=graph_trainer_llama3_1b ./run_train.sh \
      --compile.mode aot_fx_trace --compile.memory_policy default \
      --training.local_batch_size 4 --training.seq_len 2048 \
      --compile.disable_passes cudagraph_pass,tag_with_memory_policy_pass,apply_cpu_offload_pass,selective_activation_remat_pass \
      --hf_assets_path ./tests/assets/tokenizer --dataloader.dataset c4_test \
      --training.steps 5
    # qwen3:  MODULE=graph_trainer.qwen3  CONFIG=graph_trainer_qwen3_1_7b

To reproduce the whole table below, sweep that command over (model, batch, seq):

    CONFIGS=(
      "llama3 graph_trainer_llama3_1b 4 2048"
      "llama3 graph_trainer_llama3_1b 8 2048"
      "llama3 graph_trainer_llama3_1b 4 4096"
      "qwen3  graph_trainer_qwen3_1_7b 4 2048"
      "qwen3  graph_trainer_qwen3_1_7b 8 2048"
      "qwen3  graph_trainer_qwen3_1_7b 4 4096"
    )
    for cfg in "${CONFIGS[@]}"; do
      read -r mod conf bs sl <<< "$cfg"
      NGPU=1 MODULE=graph_trainer.$mod CONFIG=$conf ./run_train.sh \
        --compile.mode aot_fx_trace --compile.memory_policy default \
        --training.local_batch_size $bs --training.seq_len $sl \
        --compile.disable_passes cudagraph_pass,tag_with_memory_policy_pass,apply_cpu_offload_pass,selective_activation_remat_pass \
        --hf_assets_path ./tests/assets/tokenizer --dataloader.dataset c4_test \
        --training.steps 5 2>&1 | grep "runtime-estimation"
    done

OOM note: increase batch/seq carefully. The configs below fit on one 95 GB H100;
qwen3-1.7B and llama3-1B at bs8/seq4096 OOM the fused step on a single H100.

# results:

Runtime estimation -- the three modes vs the real fused execution, POST-regional
graph, single H100, `--compile.memory_policy default`. All times in milliseconds.
`interp(e2e)` (real overlap-aware end-to-end run) is the ground truth.

| model       | bs | seq  | tokens | cost-model | benchmark | interp(sum) | interp(e2e) | cm/e2e | bm/e2e |
|-------------|----|------|--------|------------|-----------|-------------|-------------|--------|--------|
| llama3-1B   | 4  | 2048 |  8K    | 149.978    | 167.615   | 184.156     | 177.622     | 0.844  | 0.944  |
| llama3-1B   | 8  | 2048 | 16K    | 274.058    | 290.594   | 323.049     | 320.221     | 0.856  | 0.907  |
| llama3-1B   | 4  | 4096 | 16K    | 274.062    | 290.396   | 333.833     | 330.258     | 0.830  | 0.879  |
| qwen3-1.7B  | 4  | 2048 |  8K    | 230.387    | 276.333   | 304.602     | 291.522     | 0.790  | 0.948  |
| qwen3-1.7B  | 8  | 2048 | 16K    | 428.913    | 482.989   | 528.709     | 524.662     | 0.818  | 0.921  |
| qwen3-1.7B  | 4  | 4096 | 16K    | 428.937    | 483.120   | 543.571     | 538.942     | 0.796  | 0.896  |

Observations:
  - cost-model is a under-predictor, ~0.79-0.86x of real across both
    models and all configs. A single scale factor would not work as per-op
    differences are non-negligible. Check the per-op breakdown for details.
  - benchmark is closer, ~0.88-0.95x of real; its residual gap is the
    un-benchmarkable flex HOP (falls back to the roofline) plus reshape noise.
    The per-op breakdown shows that the benchmarked ops are pretty close.
  - interp(sum) ~= interp(e2e) at every config -> these models are GPU-bound;
    negligible kernel overlap, so the serialized per-op breakdown faithfully
    decomposes the real wall-clock.

## Scope / limitations

- Single-GPU (NGPU=1); not yet validated under FSDP/TP/EP. Collectives are not
  costed (compute-only estimate).
- Dense models only (MoE out of scope: dynamic routing shapes).
- The cost-model under-counts fused-attention compute (the fused flex HOP is not
  in the flop registry); the interpreter mode captures it.
- The values are not cached yet. Working on this.

### CODE REVIEW

**Comment by yushangdi** [comment]
can you also add the tests to this PR?

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 1]
```diff
@@ -208,19 +215,91 @@ def _make_fx_forward_backward_step(
...
                 self._traced_step.gm = apply_graph_passes(
                     self._traced_step.gm,
                     self._traced_step.example_inputs,
                     passes,
                     compile_config=self.config.compile,
                 )
+            print("============== After regional inductor pass: ==============")
+            # Runtime estimation on the final (post-regional-inductor) graph.
+            measure_time(
+                _traced_step=self._traced_step,
```
Comment (yushangdi) [inline torchtitan/experiments/graph_trainer/trainer.py:296]: is this for testing only? we probably don't want to run this by default. or maybe just not land it here and only do it in local script.

Comment (nnurlan008) [reply torchtitan/experiments/graph_trainer/trainer.py:296]: I added this to trainer.py in graph_trainer to print the results. I will change this so that the prints are only enabled by a flag provided in the flag.
